site/components/src/main/java/com/visitscotland/brxm/services/FileMetaDataCalculator.java

```
@@ -0,0 +1,101 @@
+package com.visitscotland.brxm.services;
+
+import com.visitscotland.brxm.model.AssetLink;
+import com.visitscotland.brxm.model.bsh.FileMetaData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+import java.util.Locale;
+import java.util.Optional;
+
+@Component
+public class FileMetaDataCalculator {
+
+    private static final Logger logger = LoggerFactory.getLogger(FileMetaDataCalculator.class);
+
+    private final HttpConnectionProvider connectionProvider;
+
+    public FileMetaDataCalculator(HttpConnectionProvider connectionProvider) {
+        this.connectionProvider = connectionProvider;
+    }
+
+    private FileMetaData getFileMetaData(String link) {
+        HttpURLConnection connection = null;
+        try {
+            connection = connectionProvider.openConnection(link);
+            if (connection.getResponseCode() >= 400) {
+                return null;
```

I would say that this is an exceptional circumstance, should we throw an exception?

---

Our policy in this proyect is to recover from errors instead of propagating them, so we better stick with that?

I know we could get roughly the same by throwing an Exception but my concern is that the potential uses of this class are managing their way with the expected output `null` and we might need to do a big refactoring if we go down this route

---

* Clarity and Explicitness: Exceptions clearly signal that something unexpected has occurred, making it easier for developers to identify and handle critical issues. In contrast, error recovery might mask problems, leading to silent failures that are harder to diagnose.
* Separation of Concerns: Throwing exceptions allows the error-handling logic to be separated from the main code flow. This can make the code cleaner and more maintainable, as error handling is not mixed with the primary functionality.
* Flexibility in Handling: Exceptions can be caught and handled at different levels of the application, giving developers flexibility in how they manage errors. This flexibility is particularly useful in complex systems where different components might need to handle errors differently.
* Preventing Data Corruption: In some cases, continuing to execute code after an error can lead to data corruption or inconsistent states. Throwing an exception ensures that the application stops executing potentially harmful code, preserving data integrity.
* Documentation and Debugging: Exceptions often provide detailed information about what went wrong, which can be invaluable for debugging. This can lead to faster resolution of issues and better documentation of error conditions.
* User Experience: While error recovery might seem user-friendly, it can sometimes lead to confusing behavior if not properly communicated. Exceptions, when handled correctly, can provide clear feedback to users about what went wrong, helping them understand and potentially resolve the issue.

---

Even though I agree with these principles, I think still get some benefits from proper error recovery. Usually the principle in this project is to try to deliver as much value as possible to the user so, when we cannot provide some information we just ignore it because we can still deliver some valuable information.

In my head, The exception propagation and encapsulation works well when you try to perform an operation that cannot be completed for some reason. I that situation you want to feed back as much information as possible to simplify troubleshooting and that's critical.

However, this application is even more complex than a Website, as we have several types of users who would benefit from different kind of information:
- Final users: They don't care about development problems, they just want to get the data. Error pages just put them off as the site would look unstable
- CMS users: They benefit from content related feedback such as "This document you are linking to doesn't seem to be available, and we are hiding it from you". These users benefit from *preview error messages* (WIP?) and the Content Logger (WIP)
- Developers: They benefit from information about code doing stuff that is not expected. They shouldn't care about an external document not being available at a point in time.

Obviously the user that suffers the most about not having proper error propagation are the developers (closely followed by CMS Users) so let's try to pay attention to what we log

Please, don't get me wrong, I still agree with having a proper exception management, but this project might not be ideal for extensive error propagation.

---

As I was writing this, I was analysing how an exception would help in this case and even though the propagation wouldn't go far I kind of see how we can benefit from it in this case.